### DISCUSSION TITLE
EMQX Dashboard not opening on HTTPS

### DISCUSSION THREAD

**Discussion by nishant-vvdn** [emqx/emqx#4685] [Q&A]
I have used EMQX v4.2.5 and later. EMQX Dashboard is not opening on HTTPS when using in kubernetes deployment.
Is this a bug?

**Comment by zmstone** [comment]
Hi @nishant-vvdn
could you share the info how you configured `emqx_management` plugin ?

**Reply by nishant-vvdn** [reply]
Hi, @zmstone 

This is how management.conf is looking right now.

```
##--------------------------------------------------------------------
## EMQ X Management Plugin
##--------------------------------------------------------------------

## Max Row Limit
management.max_row_limit = 10000

## Application default secret
##
## Value: String
## management.application.default_secret = 

## Default Application ID
##
## Value: String
management.default_application.id = 

## Default Application Secret
##
## Value: String
management.default_application.secret = 

##--------------------------------------------------------------------
## HTTP Listener

management.listener.http = 8081
management.listener.http.acceptors = 2
management.listener.http.max_clients = 512
management.listener.http.backlog = 512
management.listener.http.send_timeout = 15s
management.listener.http.send_timeout_close = on
management.listener.http.inet6 = false
management.listener.http.ipv6_v6only = false

##--------------------------------------------------------------------
## HTTPS Listener

## management.listener.https = 8081
## management.listener.https.acceptors = 2
## management.listener.https.max_clients = 512
## management.listener.https.backlog = 512
## management.listener.https.send_timeout = 15s
## management.listener.https.send_timeout_close = on
## management.listener.https.certfile = etc/certs/cert.pem
## management.listener.https.keyfile = etc/certs/key.pem
## management.listener.https.cacertfile = etc/certs/cacert.pem
## management.listener.https.verify = verify_peer
## management.listener.https.tls_versions = tlsv1.2,tlsv1.1,tlsv1
## management.listener.https.ciphers = ECDHE-ECDSA-AES256-GCM-SHA384,ECDHE-RSA-AES256-GCM-SHA384,ECDHE-ECDSA-AES256-SHA384,ECDHE-RSA-AES256-SHA384,EC
## management.listener.https.fail_if_no_peer_cert = true
## management.listener.https.inet6 = false
## management.listener.https.ipv6_v6only = false
```

**Comment by k32** [comment]
>  is not opening on HTTPS when using in kubernetes deployment.

Could you clarify what behavior do you see, and what behavior do you expect? Do you see dashboard in plain text, but expect to see it HTTPS? According to the configuration file that you've posted above, HTTPS is not configured in EMQ X. 

So the dashboard will be available in plain text, unless you setup a front-proxy (like nginx) in front of it.

**Reply by nishant-vvdn** [reply]
Hi, @k32 

Do we not have to configure **emqx_dashboard.conf** for the HTTPS enabled Dashboard?

Because without configuring **emqx_management.conf** HTTPS Dashboard was working in **v4.0.3**. 
What changed in latest versions?

**Reply by k32** [reply]
Looking at this message (https://github.com/emqx/emqx/discussions/4685#discussioncomment-668712) the configuration for HTTPS listener is commented out.

**Reply by k32** [reply]
I don't think the defaults changed between the versions. Here's the default config file for `v4.0.3`: https://github.com/emqx/emqx-management/blob/v4.0.3/etc/emqx_management.conf HTTPS listener is also commented out

**Reply by nishant-vvdn** [reply]
So when we were not configuring **emqx_management.conf** in **v4.0.3** and HTTPS was working. It should probably work in **v4.2.5** as well.

**Reply by k32** [reply]
Can you post your `emqx_management.conf` from the `v4.0.3` installation?

**Reply by nishant-vvdn** [reply]
Since we are not configuring emqx_management.conf, it is in default state as in the v4.2.5 version.

```
##--------------------------------------------------------------------
## EMQ X Management Plugin
##--------------------------------------------------------------------

## Max Row Limit
management.max_row_limit = 10000

## Application default secret
##
## Value: String
## management.application.default_secret = 

## Default Application ID
##
## Value: String
management.default_application.id = 

## Default Application Secret
##
## Value: String
management.default_application.secret = 

##--------------------------------------------------------------------
## HTTP Listener

management.listener.http = 8081
management.listener.http.acceptors = 2
management.listener.http.max_clients = 512
management.listener.http.backlog = 512
management.listener.http.send_timeout = 15s
management.listener.http.send_timeout_close = on
management.listener.http.inet6 = false
management.listener.http.ipv6_v6only = false

##--------------------------------------------------------------------
## HTTPS Listener

## management.listener.https = 8081
## management.listener.https.acceptors = 2
## management.listener.https.max_clients = 512
## management.listener.https.backlog = 512
## management.listener.https.send_timeout = 15s
## management.listener.https.send_timeout_close = on
## management.listener.https.certfile = etc/certs/cert.pem
## management.listener.https.keyfile = etc/certs/key.pem
## management.listener.https.cacertfile = etc/certs/cacert.pem
## management.listener.https.verify = verify_peer
## management.listener.https.tls_versions = tlsv1.2,tlsv1.1,tlsv1
## management.listener.https.ciphers = 
## management.listener.https.fail_if_no_peer_cert = true
## management.listener.https.inet6 = false
## management.listener.https.ipv6_v6only = false
```

**Reply by k32** [reply]
I just ran emqx 4.0.3 docker container from the docker hub, with the default settings, and the dashboard uses plaintext connection. Please check if you changed setup of your ingress controller in k8s.

**Reply by nishant-vvdn** [reply]
No, ingress controller setup has not been changed.
We literally changed nothing except the version of EMQX.

**Reply by k32** [reply]
I went through all tags of https://github.com/emqx/emqx-dashboard/blob/v4.2.5/etc/emqx_dashboard.conf and I saw no changes since 2019. Same goes for the schema file: https://github.com/emqx/emqx-dashboard/blob/v4.2.5/priv/emqx_dashboard.schema 

In order to proceed investigation I need the following data:

1. Full hashes of EMQ X docker images used in the both `v4.0.3` and `v4.2.5` setups
2. Which environment variables do you set for the EMQ X container
2. Configuration of the ingress controller
3. Which URL do you access in the browser

**Reply by nishant-vvdn** [reply]
> In order to proceed investigation I need the following data:
> 
> 1. Full hashes of EMQ X docker images used in the both v4.0.3 and v4.2.5 setups
> 2. Which environment variables do you set for the EMQ X container
> 3. Configuration of the ingress controller
> 4. Which URL do you access in the browser

1. Image: emqx/emqx:v4.0.3
    Image ID: docker-pullable://emqx/emqx@sha256:7df415540df88e8f6945dfa8ad54a1022e48d25d5c40a61fff4d8f3850c747af
    
Image: emqx/emqx:4.2.5
Image ID: docker-pullable://emqx/emqx@sha256:1b0be91416b6e9adef1e3f6913d467c3c85cbcb1d995738cde6b2d63b26642e0

2. env:
        - name: EMQX_LOADED_PLUGINS
          value: emqx_management,emqx_recon,emqx_retainer,emqx_dashboard,emqx_rule_engine,emqx_auth_username
        - name: EMQX_CLUSTER__DISCOVERY
          value: k8s
        - name: EMQX_NAME
          value: emqx
        - name: EMQX_CLUSTER__K8S__APISERVER
          value: https://kubernetes.default.svc:443
        - name: EMQX_CLUSTER__K8S__SUFFIX
          value: svc.cluster.local
        - name: EMQX_CLUSTER__K8S__NAMESPACE
          value: emqx
        - name: EMQX_CLUSTER__K8S__SERVICE_NAME
          value: emqx
        - name: EMQX_CLUSTER__K8S__ADDRESS_TYPE
          value: hostname
        - name: EMQX_CLUSTER__K8S__APP_NAME
          value: emqx
        - name: EMQX_LISTENER__SSL__EXTERNAL__KEYFILE
          value: 
        - name: EMQX_LISTENER__SSL__EXTERNAL__CERTFILE
          value: 
        - name: EMQX_LISTENER__SSL__EXTERNAL__CACERTFILE
          value:
        - name: EMQX_LISTENER__SSL__EXTERNAL__VERIFY
          value: verify_peer
        - name: EMQX_ACL_NOMATCH
          value: "deny"
        - name: EMQX_ACL_DENY_ACTION
          value: "disconnect"
        - name: EMQX_LISTENER__SSL__EXTERNAL__FAIL_IF_NO_PEER_CERT
          value: "true"
        - name: EMQX_ALLOW_ANONYMOUS
          value: "false"
        - name: EMQX_LISTENER__SSL__EXTERNAL__MAX_CONNECTIONS
          value: "1024000"
        - name: EMQX_AUTH__USER__PASSWORD_HASH
          value: sha256
        - name: EMQX_AUTH__USER__1__USERNAME
          value:
        - name: EMQX_AUTH__USER__1__PASSWORD
          value:
        - name: EMQX_DASHBOARD__DEFAULT_USER__LOGIN
          value:
        - name: EMQX_DASHBOARD__DEFAULT_USER__PASSWORD
          value:
        - name: EMQX_DASHBOARD__LISTENER__HTTPS
          value: "18084"
        - name: EMQX_DASHBOARD__LISTENER__HTTPS__ACCEPTORS
          value: "4"
        - name: EMQX_DASHBOARD__LISTENER__HTTPS__MAX_CLIENTS
          value: "512"
        - name: EMQX_DASHBOARD__LISTENER__HTTPS__CERTFILE
          value:
        - name: EMQX_DASHBOARD__LISTENER__HTTPS__KEYFILE
          value:

3. We are **not using ingress controller** but the internal **Kubernetes service** is working as a **NodeBalancer**.

4. URL is our internal DNS mapping which unfortunately we cannot share.

**Reply by k32** [reply]
Thanks for the info. I am not very familiar with Kubernetes, but from my attempts to reproduce this issue in docker:

```
$ docker run -p 18084:18084 -e EMQX_DASHBOARD__LISTENER__HTTPS=18084 -e EMQX_DASHBOARD__LISTENER__HTTPS__CERTFILE=etc/certs/cert.pem -e EMQX_DASHBOARD__LISTENER__HTTPS__KEYFILE=etc/certs/key.pem -e EMQX_DASHBOARD__LISTENER__HTTPS__MAX_CLIENTS=200 -e EMQX_DASHBOARD__LISTENER__HTTPS__ACCEPTORS=5 -it emqx/emqx:4.2.5

<Output>
dashboard.listener.https=18084
dashboard.listener.https.acceptors=5
dashboard.listener.https.certfile=etc/certs/cert.pem
dashboard.listener.https.keyfile=etc/certs/key.pem
dashboard.listener.https.max_clients=200
```

I see that the dashboard starts listening on HTTPS. So the mapping between the environment variable and the configuration value didn't change at least. Could you please share the logs of EMQ X container?

**Reply by nishant-vvdn** [reply]
`2021-04-30 11:37:04.496 [error] Ranch listener 'https:dashboard' had connection process started with cowboy_tls:start_link/4 at <0.27251.621> exit with reason: {undef,[{cow_http2_machine,get_connection_local_buffer_size,[{http2_machine,server,#{},normal,undefined,#Ref<0.1636791437.2178416641.87729>,#{initial_window_size => 65535},#{},#{header_table_size => 65536,initial_window_size => 6291456,max_concurrent_streams => 1000,max_header_list_size => 262144},15728640,65535,2,1,[{stream,1,<<"GET">>,nofin,6291456,{[],[]},0,undefined,fin,65535,0,0,undefined}],[],[],{state,1065,4096,4096,[{68,{<<"cookie">>,<<"_ga=GA1.2.785654357.1615357055">>}},{73,{<<"accept-language">>,<<"en-GB,en-US;q=0.9,en;q=0.8">>}},{64,{<<"accept-encoding">>,<<"gzip, deflate, br">>}},{54,{<<"sec-fetch-dest">>,<<"document">>}},{48,{<<"sec-fetch-user">>,<<"?1">>}},{54,{<<"sec-fetch-mode">>,<<"navigate">>}},{50,{<<"sec-fetch-site">>,<<"none">>}},{173,{<<"accept">>,<<"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,/;q=0.8,application/signed-exchange;v=b3;q=0.9">>}},{146,{<<"user-agent">>,<<"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36">>}},{58,{<<"upgrade-insecure-requests">>,<<"1">>}},{50,{<<"sec-ch-ua-mobile">>,<<"?0">>}},{105,{<<"sec-ch-ua">>,<<"\"GoogleChrome\";v=\"89\", \"Chromium\";v=\"89\", \";Not A Brand\";v=\"99\"">>}},{54,{<<"cache-control">>,<<"max-age=0">>}},{68,{<<":authority">>,<<"INTERNAL_DNS">>}}]},{state,384,4096,4096,[{44,{<<"server">>,<<"Cowboy">>}},{74,{<<"last-modified">>,<<"Mon, 28 Dec 2020 05:29:58 GMT">>}},{48,{<<"etag">>,<<"\"2953066607\"">>}},{65,{<<"date">>,<<"Fri, 30 Apr 2021 11:37:03 GMT">>}},{53,{<<"content-type">>,<<"text/html">>}},{50,{<<"content-length">>,<<"1130">>}},{50,{<<"accept-ranges">>,<<"bytes">>}}]}}],[]},{cowboy_http2,maybe_send_data_alarm,3,[{file,"cowboy_http2.erl"},{line,823}]},{cowboy_http2,commands,3,[{file,"cowboy_http2.erl"},{line,622}]},{cowboy_http2,loop,2,[{file,"cowboy_http2.erl"},{line,247}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}`

This log comes up when we try to access EMQX HTTPS Dashboard.
**Note: HTTPS Dashboard does load successfully but the page is blank. Just a black blank webpage.**

**Reply by k32** [reply]
This looks https://github.com/emqx/emqx/issues/4047 . I reopened it.

**Reply by nishant-vvdn** [reply]
Thanks, @k32 

I'll be looking out for a new release, then.

**Comment by nishant-vvdn** [comment]
Hi, @k32  and @zmstone 

Is this issue fixed in the latest release?

**Reply by k32** [reply]
4.3.0 includes the fix

**Reply by nishant-vvdn** [reply]
@k32 Working now, thanks!
